rust/worker/src/execution/operators/prefetch.rs

```diff
@@ -0,0 +1,196 @@
+use thiserror::Error;
+use tonic::async_trait;
+
+use crate::{
+    blockstore::provider::BlockfileProvider,
+    errors::{ChromaError, ErrorCodes},
+    execution::operator::Operator,
+    segment::record_segment::RecordSegmentReader,
+    types::Segment,
+};
+
+#[derive(Debug)]
+pub(crate) struct OffsetIdToDataKeys {
+    pub(crate) keys: Vec<u32>,
+}
+
+#[derive(Debug)]
+pub(crate) struct UserIdToOffsetIdKeys {
+    pub(crate) keys: Vec<String>,
+}
+
+#[derive(Debug)]
+pub(crate) struct OffsetIdToUserIdKeys {
+    pub(crate) keys: Vec<u32>,
+}
+
+#[derive(Debug)]
+pub(crate) enum RecordSegmentKeys {
+    OffsetIdToDataKeys(OffsetIdToDataKeys),
+    UserIdToOffsetIdKeys(UserIdToOffsetIdKeys),
+    OffsetIdToUserIdKeys(OffsetIdToUserIdKeys),
+}
+
+#[derive(Debug)]
+pub(crate) struct StringMetadataKeys {
+    pub(crate) prefixes: Vec<String>,
+    pub(crate) keys: Vec<String>,
+}
+
+#[derive(Debug)]
+pub(crate) struct F32MetadataKeys {
+    pub(crate) prefixes: Vec<String>,
+    pub(crate) keys: Vec<String>,
+}
+
+#[derive(Debug)]
+pub(crate) struct BoolMetadataKeys {
+    pub(crate) prefixes: Vec<String>,
+    pub(crate) keys: Vec<String>,
+}
+
+#[derive(Debug)]
+pub(crate) struct U32MetadataKeys {
+    pub(crate) prefixes: Vec<String>,
+    pub(crate) keys: Vec<String>,
+}
+
+#[derive(Debug)]
+pub(crate) enum MetadataKeys {
+    StringMetadataKeys(StringMetadataKeys),
+    F32MetadataKeys(F32MetadataKeys),
+    BoolMetadataKeys(BoolMetadataKeys),
+    U32MetadataKeys(U32MetadataKeys),
+}
+
+#[derive(Debug)]
+pub(crate) struct PostingsListKeys {
+    pub(crate) prefixes: Vec<String>,
+    // Doc Id.
+    pub(crate) keys: Vec<u32>,
+}
+
+#[derive(Debug)]
+pub(crate) struct FrequenciesKeys {
+    pub(crate) prefixes: Vec<String>,
+    // Frequency.
+    pub(crate) keys: Vec<u32>,
+}
+
+#[derive(Debug)]
+pub(crate) enum FullTextKeys {
+    PostingsListKeys(PostingsListKeys),
+    FrequenciesKeys(FrequenciesKeys),
+}
+
+#[derive(Debug)]
+pub(crate) enum MetadataSegmentKeys {
+    MetadataKeys(MetadataKeys),
+    FullTextKeys(FullTextKeys),
+}
+
+#[derive(Debug)]
+pub(crate) struct RecordSegmentInfo {
+    pub(crate) keys: RecordSegmentKeys,
+    pub(crate) segment: Segment,
+    pub(crate) provider: BlockfileProvider,
+}
+
+#[derive(Debug)]
+pub(crate) struct MetadataSegmentInfo {
+    keys: MetadataSegmentKeys,
+    segment: Segment,
+    provider: BlockfileProvider,
+}
+
+#[derive(Debug)]
+pub(crate) struct HnswSegmentInfo {
+    // TODO.
+}
+
+#[derive(Debug)]
+pub(crate) enum PrefetchData {
```

Comment:
Would prefer if this was not one operator but separate operators per input type, this feels very coarse

Reply:
Done